test/integration/apiserver/flowcontrol/fight_test.go

```diff
@@ -0,0 +1,192 @@
+
+   This is a behavioral test: it instantiates these controllers and runs them
+   almost normally.  The test aims to run the controllers for a little under
+   2 minutes.  The test takes clock readings to get upper and lower bounds on
+   how long each controller ran, and calculates consequent bounds on the number
+   of writes that should happen to each FlowSchemaStatus.  The test creates
+   an informer to observe the writes.  The calculated lower bound on the
+   number of writes is very lax, assuming only that one write can be done
+   every 10 seconds.
+*/
+type fightTest struct {
+	t              *testing.T
+	ctx            context.Context
+	loopbackConfig *rest.Config
+	teamSize       int
+	stopCh         chan struct{}
+	now            time.Time
+	clk            *clock.FakeClock
+	ctlrs          map[bool][]utilfc.Interface
+
+	countsMutex sync.Mutex
+
+	// writeCounts maps FlowSchema.Name to number of writes
+	writeCounts map[string]int
+}
+
+func newFightTest(t *testing.T, loopbackConfig *rest.Config, teamSize int) *fightTest {
+	now := time.Now()
+	ft := &fightTest{
+		t:              t,
+		ctx:            context.Background(),
+		loopbackConfig: loopbackConfig,
+		teamSize:       teamSize,
+		stopCh:         make(chan struct{}),
+		now:            now,
+		clk:            clock.NewFakeClock(now),
+		ctlrs: map[bool][]utilfc.Interface{
+			false: make([]utilfc.Interface, teamSize),
+			true:  make([]utilfc.Interface, teamSize)},
+		writeCounts: map[string]int{},
+	}
+	return ft
+}
+
+func (ft *fightTest) createMainInformer() {
+	myConfig := rest.CopyConfig(ft.loopbackConfig)
+	myConfig = rest.AddUserAgent(myConfig, "audience")
+	myClientset := clientset.NewForConfigOrDie(myConfig)
+	informerFactory := informers.NewSharedInformerFactory(myClientset, 0)
+	inf := informerFactory.Flowcontrol().V1beta1().FlowSchemas().Informer()
+	inf.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			fs := obj.(*flowcontrol.FlowSchema)
+			ft.countWrite(fs)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			fs := newObj.(*flowcontrol.FlowSchema)
+			ft.countWrite(fs)
+		},
+	})
+	go inf.Run(ft.stopCh)
+	if !cache.WaitForCacheSync(ft.stopCh, inf.HasSynced) {
+		ft.t.Errorf("Failed to sync main informer cache")
+	}
+}
+
+func (ft *fightTest) countWrite(fs *flowcontrol.FlowSchema) {
+	ft.countsMutex.Lock()
+	defer ft.countsMutex.Unlock()
+	ft.writeCounts[fs.Name]++
+}
+
+func (ft *fightTest) createController(invert bool, i int) {
+	fieldMgr := fmt.Sprintf("testController%d%v", i, invert)
+	myConfig := rest.CopyConfig(ft.loopbackConfig)
+	myConfig = rest.AddUserAgent(myConfig, fieldMgr)
+	myClientset := clientset.NewForConfigOrDie(myConfig)
+	fcIfc := myClientset.FlowcontrolV1beta1()
+	informerFactory := informers.NewSharedInformerFactory(myClientset, 0)
+	foundToDangling := func(found bool) bool { return !found }
+	if invert {
+		foundToDangling = func(found bool) bool { return found }
+	}
```

Comment:
nit: you could replace these 4 lines with

foundToDangling := func(found bool) bool { return invert == dangling }

Reply:
@wojtek-t : good catch.

@deads2k : yes, this is to make the two teams of controllers fight.  I do not understand the suggestion about two sets of data.  We need a disagreement between controllers over the same data.